code/go/0chain.net/smartcontract/storagesc/challenge.go

```diff
@@ -667,48 +777,43 @@ func (sc *StorageSmartContract) generateChallenges(t *transaction.Transaction,
 	return nil
 }
 
-func (sc *StorageSmartContract) addChallenge(alloc *StorageAllocation,
+func (sc *StorageSmartContract) addChallenge(alloc *StorageAllocation, blobberID string,
 	validators partitions.RandPartition, challengeID string,
 	creationDate common.Timestamp, r *rand.Rand, challengeSeed int64,
 	balances c_state.StateContextI) (resp string, err error) {
 
-	sort.SliceStable(alloc.Blobbers, func(i, j int) bool {
-		return alloc.Blobbers[i].ID < alloc.Blobbers[j].ID
-	})
-
-	selectedBlobberObj := &StorageNode{}
-
-	blobberAllocation := &BlobberAllocation{}
-	blobberAllocation.Stats = &StorageAllocationStats{}
+	if blobberID == "" {
+		return "", common.NewError("add_challenges",
+			"empty blobber id")
+	}
+	blobber := &StorageNode{}
 
-	for _, ri := range r.Perm(len(alloc.Blobbers)) {
-		selectedBlobberObj = alloc.Blobbers[ri]
-		_, ok := alloc.BlobberMap[selectedBlobberObj.ID]
-		if !ok {
-			Logger.Error("Selected blobber not found in allocation state",
-				zap.Any("selected_blobber", selectedBlobberObj),
-				zap.Any("blobber_map", alloc.BlobberMap))
-			return "", common.NewError("invalid_parameters",
-				"Blobber is not part of the allocation. Could not find blobber")
-		}
-		blobberAllocation = alloc.BlobberMap[selectedBlobberObj.ID]
-		if blobberAllocation.AllocationRoot != "" {
-			break // found
+	for _, b := range alloc.Blobbers {
+		if b.ID == blobberID {
+			blobber = b
+			break
 		}
 	}
+	if blobber == nil {
```

Comment:
blobber is never nil as set to `&StorageNode{}` above. Check `blobber.ID == ""` or similar instead.